Allow marking of earlier offsets

jerrygb opened the pull request:
This commit is in alignment with the sarama library commits to reset offsets
https://github.com/Shopify/sarama/commit/b966238f31778386bce80c28a12fde729e01cab4#diff-3ca6d659defd100fe2de43adf2b8f41e
https://github.com/Shopify/sarama/commit/96fa1c837819761fc5c706010a9677aab15ca46d#diff-3ca6d659defd100fe2de43adf2b8f41e

jerrygb commented:
I believe it might be a good idea to combine these with changes from @aravindvs

dim commented on consumer.go:
```diff
@@ -161,6 +161,34 @@ func (c *Consumer) MarkOffsets(s *OffsetStash) {
+//
+// Difference between ResetOffset and MarkOffset is that it allows to rewind to an earlier offset
+func (c *Consumer) ResetOffset(msg *sarama.ConsumerMessage, metadata string) {
+	c.subs.Fetch(msg.Topic, msg.Partition).MarkOffset(msg.Offset+1, metadata)
```
Shouldn't that call `.ResetOffset(...)`

jerrygb commented:
@dim I have fixed as per the comment. You were right. Additionally, added some tests.

jerrygb commented:
@dim reopening for travis rebuild, due to seemingly okay test
https://travis-ci.org/bsm/sarama-cluster/jobs/322296975

dim commented on cluster_test.go:
```diff
@@ -21,6 +21,7 @@ var (
 	testKafkaRoot  = "kafka_2.12-1.0.0"
 	testKafkaAddrs = []string{"127.0.0.1:29092"}
 	testTopics     = []string{"topic-a", "topic-b"}
+	testTopicsReset     = "topic-c"
```
pls use `gofmt` or `goimports`

dim replied:
also, no real need to declare this as global, it's only used once

jerrygb replied:
True. Removed assuming we dont need another fuzzing test.

dim commented on cluster_test.go:
```diff
@@ -73,6 +74,10 @@ var _ = BeforeSuite(func() {
 		"-name", "kafkaServer", "kafka.Kafka",
 		testDataDir("server.properties"),
 	)
+	if _, err := os.Stat(testKafkaData); err == nil  {
```
No need for that clause, instead just the below will do:

```
Expect(os.RemoveAll(testKafkaData)).To(Succeed())
```

dim commented on consumer_test.go:
```diff
@@ -315,6 +340,94 @@ var _ = Describe("Consumer", func() {
 		Expect(uniques).To(HaveLen(15000))
 	})
 
+	It("should consume/commit/reset/resume", func() {
```
so I am not really sure we need another fuzzing test for this. I don't see the added value. what you have done in https://github.com/bsm/sarama-cluster/pull/197/files#diff-e85d03976d9766c4d00ac176511d4f06R188 seems enough, please remove that test.

jerrygb replied:
Removed the test. This was meant to check that we do actually rewind and the first offset is actually the one we expect. The earlier fuzzing test actually doesnt deal specifically with start offsets, right? 

Rather it deals with consuming atleast fixed number of unique entries (partition/offset pairs)?

I have removed it as you advised. But, let me know if we need check the start offsets when we start consumption.

jerrygb commented:
@dim fixed as per comments. 

Also removed the additional fuzzing test. This was meant to check that we do actually rewind and, whether the first offset consumed is actually the one we expected. I suppose the earlier  fuzzing test actually deal specifically with start offsets, right?

For now, I have removed it as you advised. But, let me know if you need to me add those back in.
